loki/backend/cgen.py

```diff
@@ -140,6 +143,25 @@ def map_c_reference(self, expr, enclosing_prec, *args, **kwargs):
     def map_c_dereference(self, expr, enclosing_prec, *args, **kwargs):
         return self.format(' (*%s)', self.rec(expr.expression, PREC_NONE, *args, **kwargs))
 
+    def map_inline_call(self, expr, enclosing_prec, *args, **kwargs):
+
+        class FindFloatLiterals(ExpressionFinder):
+            retriever = ExpressionRetriever(lambda e: isinstance(e, sym.FloatLiteral))
+
+        if expr.function.name.lower() == 'mod':
+            parameters = [self.rec(param, PREC_NONE, *args, **kwargs) for param in expr.parameters]
+            # TODO: this check is not quite correct, as it should evaluate the
+            #  expression(s) of both arguments/parameters and choose the integer version of modulo ('%')
+            #  instead of the floating-point version ('fmod')
+            #  whenever the mentioned evaluations result in being of kind 'integer' ...
+            #  as an example: 'celing(3.1415)' got an floating point value in it, however it evaluates/returns
+            #  an integer, in that case the wrong modulo function/operation is chosen
```

Comment:
[no action] This is indeed a tricky corner case. Solution looks good for now, but might need some more thought when this gets problematic.

Reply:
Likely the correct approach would be an `ExpressionTypeMapper` that attempts to determine the return type of an expression (similar to [`ExpressionDimensionsMapper`](https://github.com/ecmwf-ifs/loki/blob/cc179b3b86613af86df6dabb0be250cde4414fc9/loki/expression/mappers.py#L362)) - but it will require baking in a lot of knowledge about Fortran intrinsics. Fparser might provide some of that, so it could be feasible.

Definitely way beyond the scope of this PR, though.

Reply:
Yes! We could also think about implementing a C++ (templated) function `mymod()` which either calls `fmod` or uses the modulo operator in dependence of the arguments types.